### PR TITLE
DPL: adapt filters to be used over joins and concats

### PR DESCRIPTION
**Pull request by aalkin**
Streamlines the manipulation of filtered/join/concat by tracking the types of predefined tables so they can be bound from external inputs.

### CODE REVIEW

**Comment by aalkin** [comment]
@ktf this is what we intended, however it will not work just yet, as tracking of originals is incorrect for RowView - it uses RowViewBase::table_t for iterators from joins/concats

File: Framework/Core/include/Framework/AnalysisTask.h [hunk 1]
```diff
@@ -405,13 +391,16 @@ struct AnalysisDataProcessorBuilder {
...
+          if constexpr (is_specialization<std::decay_t<AssociatedType>, o2::soa::Table>::value) {
             for (auto& groupedDatum : groupsCollection) {
               auto groupedElementsTable = arrow::util::get<std::shared_ptr<arrow::Table>>(groupedDatum.value);
               task.process(groupingElement, AssociatedType{groupedElementsTable, offsets[oi]});
               ++const_cast<std::decay_t<Grouping>&>(groupingElement);
               ++oi;
             }
+          }
+          if constexpr (is_specialization<std::decay_t<AssociatedType>, o2::soa::Filtered>::value) {
+            // FIXME: we need to implement the case for the grouped filtered case.
```
Comment (ktf) [inline Framework/Core/include/Framework/AnalysisTask.h]: Maybe a static_assert?

File: Framework/Core/include/Framework/AnalysisTask.h [hunk 2]
```diff
@@ -613,13 +602,15 @@ DataProcessorSpec adaptAnalysisTask(std::string name, Args&&... args)
                 "At least one of process(...), T::run(...), init(...) must be defined");
 
   std::vector<InputSpec> inputs;
-  std::vector<SchemaInfo> filteredSchemas;
+  // vector of pairs (number of argument in process(), pointer to gandiva expression tree)
```
Comment (ktf) [inline Framework/Core/include/Framework/AnalysisTask.h]: How about making it an actual struct?

File: Framework/Core/include/Framework/Expressions.h [hunk 3]
```diff
@@ -23,8 +23,11 @@
 #include <memory>
 
 using atype = arrow::Type;
-using SchemaInfo = std::pair<std::string, gandiva::SchemaPtr>;
-using ExpressionInfo = std::pair<std::string, gandiva::NodePtr>;
+struct ExpressionInfo {
+  size_t I;
```
Comment (ktf) [inline Framework/Core/include/Framework/Expressions.h]: Why capital i? Maybe `index`?

File: Framework/Core/src/ASoA.cxx [hunk 4]
```diff
@@ -16,6 +16,8 @@ namespace o2::soa
 
 std::shared_ptr<arrow::Table> ArrowHelpers::joinTables(std::vector<std::shared_ptr<arrow::Table>>&& tables)
 {
+  if (tables.size() == 1)
```
Comment (ktf) [inline Framework/Core/src/ASoA.cxx]: I think the coding conventions expect `{}`  always for `if`..

**Pull request merged**